test/matrix_test.go

```diff
@@ -0,0 +1,497 @@
+  - status: "True"
+    type: Succeeded
+`, "pr-pt1-with-result", namespace))
+
+	taskrun2withresults := parse.MustParseV1TaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  results:
+  - name: versions
+    value:
+     - go1.17
+     - go1.18.1
+  serviceAccountName: default
+  timeout: 1h0m0s
+  taskRef:
+    name: task2withresults
+    kind: Task
+status:
+  conditions:
+  - status: "True"
+    type: Succeeded
+`, "pr-pt2-with-result", namespace))
+
+	expectedTaskRuns := []*v1.TaskRun{taskRun1, taskRun2, taskRun3, taskRun4, taskRun5, taskrun1withresults, taskrun2withresults}
+
+	t.Logf("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+	if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunSucceed(prName), "PipelineRunSuccess", v1Version); err != nil {
+		t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+	}
+
+	actualTaskrunList, err := c.V1TaskRunClient.List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("tekton.dev/pipelineRun=%s", prName)})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prName, err)
+	}
+
+	for i := range actualTaskrunList.Items {
+		expectedTaskRun := expectedTaskRuns[i]
+		if d := cmp.Diff(expectedTaskRun.Spec, actualTaskrunList.Items[i].Spec, ignoreTypeMeta); d != "" {
+			t.Errorf("expected to see TaskRun %v created. Diff %s", expectedTaskRuns[i].Name, diff.PrintWantGot(d))
+		}
+	}
+
+	t.Logf("Successfully finished test TestPipelineRunMatrixed")
+}
+
+// TestPipelineRunMatrixedFailed is an integration test with a Matrixed PipelineRun
+// that contains 2 successful, and 1 unsuccessful TaskRun. This test verifies that the
+// 3rd task run failing will cause the PipelineRun to fail, however the first 2 TaskRuns
+// should succeed before the last one fails.
+func TestPipelineRunMatrixedFailed(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	c, namespace := setup(ctx, t, requireAlphaFeatureFlag)
+	knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+	defer tearDown(ctx, t, c, namespace)
+	t.Logf("Creating Task in namespace %s", namespace)
+	task := parse.MustParseV1Task(t, fmt.Sprintf(`
+metadata:
+  name: mytask
+  namespace: %s
+spec:
+  params:
+    - name: exit-code
+  steps:
+    - name: echo
+      image: alpine
+      script: |
+        exit "$(params.exit-code)"
+`, namespace))
+
+	if _, err := c.V1TaskClient.Create(ctx, task, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task `%s`: %s", task.Name, err)
+	}
+
+	pipelineRun := parse.MustParseV1PipelineRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: exit-with-zero-and-non-zero
+        taskRef:
+          name: mytask
+        matrix:
+          params:
+          - name: exit-code
+            value:
+              - "0"
```

Comment:
nit: do we need 2 taskruns w/ exit code 0, or can we get sufficient test coverage with just 1?